src/main/java/space/obminyashka/items_exchange/service/impl/AdvertisementServiceImpl.java

```diff
@@ -104,17 +109,24 @@ public Optional<AdvertisementDisplayDto> findDtoById(UUID id) {
     }
 
     @Override
-    public List<AdvertisementTitleDto> findFirst10ByFilter(AdvertisementFilterDto dto) {
-        return advertisementRepository.findFirst10ByParams(
-                        dto.getAge(),
-                        dto.getGender(),
-                        dto.getSize(),
-                        dto.getSeason(),
-                        dto.getSubcategoryId(),
-                        dto.getCategoryId(),
-                        dto.getLocationId()).stream()
-                .map(this::buildAdvertisementTitle)
-                .toList();
+    public Page<AdvertisementTitleDto> findAdvertisementByFilter(AdvertisementFilterRequest dto) {
+        QAdvertisement qAdvertisement = QAdvertisement.advertisement;
+        BooleanBuilder predicate = new BooleanBuilder()
+                .and(!dto.getSubcategorySearchRequest().getSubcategoriesIdValues().isEmpty()
+                        ? qAdvertisement.subcategory.id.in(dto.getSubcategorySearchRequest().getSubcategoriesIdValues()) : null)
+                .and(!dto.getAdvertisementFilter().getClothingSizes().isEmpty()
+                        ? qAdvertisement.size.in(dto.getAdvertisementFilter().getClothingSizes().stream().map(x -> x.range)
+                        .collect(Collectors.toSet())) : null)
+                .and(!dto.getAdvertisementFilter().getShoesSizes().isEmpty() ? qAdvertisement.size.in(dto.getAdvertisementFilter().getShoesSizes().stream()
+                        .map(x -> String.valueOf(x.length))
+                        .collect(Collectors.toSet())) : null)
+                .and(!dto.getAdvertisementFilter().getSeason().isEmpty() ? qAdvertisement.season.in(dto.getAdvertisementFilter().getSeason()) : null)
+                .and(!dto.getAdvertisementFilter().getAge().isEmpty() ? qAdvertisement.age.in(dto.getAdvertisementFilter().getAge()) : null)
+                .and(dto.getAdvertisementFilter().getLocationId() != null ? qAdvertisement.location.id.eq(dto.getAdvertisementFilter().getLocationId()) : null)
+                .and(dto.getAdvertisementFilter().getGender() != null ? qAdvertisement.gender.in(dto.getAdvertisementFilter().getGender()) : null);
```

Comment:
you can remove checks on Null or Empty because QueryDsl do it automatically

Reply:
And I think Creation of predicate can be method of class AdvertisementFilterRequest with name like toPredicate()